cli/internal/cmd/upgradeapply.go

```diff
@@ -301,52 +328,39 @@ func (u *upgradeApplyCmd) migrateTerraform(cmd *cobra.Command, conf *config.Conf
 		if !flags.yes {
 			ok, err := askToConfirm(cmd, "Do you want to apply the Terraform migrations?")
 			if err != nil {
-				return res, fmt.Errorf("asking for confirmation: %w", err)
+				return state.Infrastructure{}, fmt.Errorf("asking for confirmation: %w", err)
 			}
 			if !ok {
 				cmd.Println("Aborting upgrade.")
 				// User doesn't expect to see any changes in his workspace after aborting an "upgrade apply",
 				// therefore, roll back to the backed up state.
 				if err := u.clusterUpgrader.RestoreClusterWorkspace(); err != nil {
-					return res, fmt.Errorf(
+					return state.Infrastructure{}, fmt.Errorf(
 						"restoring Terraform workspace: %w, restore the Terraform workspace manually from %s ",
 						err,
 						filepath.Join(upgradeDir, constants.TerraformUpgradeBackupDir),
 					)
 				}
-				return res, fmt.Errorf("cluster upgrade aborted by user")
+				return state.Infrastructure{}, fmt.Errorf("cluster upgrade aborted by user")
 			}
 		}
 		u.log.Debugf("Applying Terraform migrations")
+
 		infraState, err := u.clusterUpgrader.ApplyClusterUpgrade(cmd.Context(), conf.GetProvider())
 		if err != nil {
-			return infraState, fmt.Errorf("applying terraform migrations: %w", err)
-		}
-
-		// Apply possible updates to cluster ID file
-		if err := updateClusterIDFile(infraState, u.fileHandler); err != nil {
-			return infraState, fmt.Errorf("merging cluster ID files: %w", err)
+			return state.Infrastructure{}, fmt.Errorf("applying terraform migrations: %w", err)
 		}
 
-		cmd.Printf("Terraform migrations applied successfully and output written to: %s\n"+
+		cmd.Printf("Infrastructure migrations applied successfully and output written to: %s\n"+
 			"A backup of the pre-upgrade state has been written to: %s\n",
-			flags.pf.PrefixPrintablePath(constants.ClusterIDsFilename),
+			flags.pf.PrefixPrintablePath(constants.StateFilename),
 			flags.pf.PrefixPrintablePath(filepath.Join(upgradeDir, constants.TerraformUpgradeBackupDir)),
 		)
-	} else {
-		u.log.Debugf("No Terraform diff detected")
+		return infraState, nil
 	}
+
 	u.log.Debugf("No Terraform diff detected")
-	infraState, err := u.clusterShower.ShowInfrastructure(cmd.Context(), conf.GetProvider())
-	if err != nil {
-		return infraState, fmt.Errorf("getting Terraform output: %w", err)
-	}
-	state := state.NewState(infraState)
-	// TODO(elchead): AB#3424 move this to updateClusterIDFile and correctly handle existing state when writing state
-	if err := u.fileHandler.WriteYAML(constants.StateFilename, state, file.OptOverwrite); err != nil {
-		return infraState, fmt.Errorf("writing state file: %w", err)
-	}
-	return infraState, nil
```

Comment:
Is it safe to remove this?
Afaik we still need to check if a state file is missing the infrastructure state and otherwise add it.
Though this could also be done outside of this function, at the start of `upgrade apply` as part of the migration to v2.12

Reply:
How would one get into a case where a state file exists, but no infrastructure is there?

- For new clusters, the state file should be generated.
- For pre-state-file clusters, the migration should happen from id-file to state-file.

Reply:
But the `NewFromIDFile` does not set all relevant infrastructure values for the state file, does it?
We still need to supplement some from the Terraform output, which currently does not happen if that phase is skipped or there are not Terrraform changes to be made.

Reply:
I may misunderstand, but merging postMigrationInfraState into the state-file should do this IMO, as it is either populated with the values from the Terraform configuration in case of a manual migration, or the outputs of the Terraform migration in case of the "automatic" migration. Or do I miss something here?
https://github.com/edgelesssys/constellation/blob/a631138f2819bdea812fd62a83fae1afe35de19b/cli/internal/cmd/upgradeapply.go#L211-L238

Reply:
If `migrateTerraform` detects no diff, the function returns an empty Infrastructure struct, meaning this part wouldn't add any missing values to the state file.